### PR TITLE
feat(e2e): preserve browser state across journey milestones

### PR DESCRIPTION
**Pull request by leslie-heinzen**
## Summary

Explicitly selected path and journey packages now run their milestones in one Playwright browser context. Unsaved browser state can survive milestone boundaries while each milestone still produces its existing report.

The public report contract remains unchanged: each milestone still produces one `E2ETestReport` in `MultiGuideReport.reports[]`, in execution-plan order.

This is a runner-only change. It reuses the installed plugin’s existing `bundled:e2e-test` contract and does not require a Pathfinder plugin release.

Standalone guides, file-list runs, and repository sweeps keep their current isolated execution.

## What to look at

- [CLI orchestration](https://github.com/grafana/grafana-pathfinder-app/blob/ba803422c8dc23eeda409a6cc8008732b64ce63a/src/cli/commands/e2e.ts) routes only explicit path and journey selections through the shared runner. It preserves the existing blocked dependency set and `SKIPPED_PREREQ` behavior.
- [Process transport and recovery](https://github.com/grafana/grafana-pathfinder-app/blob/ba803422c8dc23eeda409a6cc8008732b64ce63a/src/cli/e2e/playwright-runner.ts) validate chain input, reconcile partial results, and keep nonzero Playwright exits visible.
- [Milestone execution](https://github.com/grafana/grafana-pathfinder-app/blob/ba803422c8dc23eeda409a6cc8008732b64ce63a/tests/e2e-runner/shared-guide-runner.spec.ts) uses one test, page, and browser context. [Guide loading](https://github.com/grafana/grafana-pathfinder-app/blob/ba803422c8dc23eeda409a6cc8008732b64ce63a/tests/e2e-runner/utils/guide-runner/run-guide.ts) replaces Pathfinder content without reloading later milestones.
- [Shared-chain semantics](https://github.com/grafana/grafana-pathfinder-app/blob/ba803422c8dc23eeda409a6cc8008732b64ce63a/tests/e2e-runner/utils/shared-chain.ts) own milestone transitions, dependency skips, session aborts, and incremental publication.
- [Session checks](https://github.com/grafana/grafana-pathfinder-app/blob/ba803422c8dc23eeda409a6cc8008732b64ce63a/tests/e2e-runner/auth/grafana-auth.ts) distinguish authentication expiry from network, server, page, context, and browser failures.
- [Milestone replacement](https://github.com/grafana/grafana-pathfinder-app/blob/ba803422c8dc23eeda409a6cc8008732b64ce63a/tests/e2e-runner/utils/guide-runner/milestone-replacement.ts) uses the existing Reset guide and `bundled:e2e-test` contracts. It clears Pathfinder progress without reloading or clearing Grafana application state.

## Why

Recent pool results found 18 learning journeys whose failed milestone saw the home page instead of its expected state. Eleven had a non-root starting location, but URL restoration could not reconstruct unsaved application state.

A learning journey represents one continuous user flow. Rewriting each milestone to reconstruct every wizard, form, and in-memory state would move runner limitations into guide content. A shared browser session tests the flow that a user actually follows.

## How to verify

1. Start the local Grafana test stack and build the CLI.
2. Run `GRAFANA_URL=http://localhost:3000 bash tests/e2e-runner/fixtures/shared-session-path/verify.sh`.
3. Confirm that the two-milestone path passes and the second milestone sees the unsaved configuration value.
4. Confirm that the same second milestone fails when the CLI runs it independently.
5. Run `create-availability-slo-lj` against an isolated development stack. Confirm that `configure-targets` remains in the live SLO wizard session.
6. Inspect the multi-guide JSON report. Confirm that it contains one ordered `reports[]` entry for every milestone.

## Breaking changes

There is no JSON report schema change. The schema remains `1.0.0`.

The installed Pathfinder plugin contract is unchanged. Existing target stacks can use the shared runner without waiting for a plugin release.

Explicit path and journey runs now preserve browser state between milestones. A caller that relied on a new browser context for each milestone will observe different runtime behavior.

## Reversibility

A revert restores per-milestone browser isolation. This change has no persisted data migration. The E2E platform can also roll back by restoring its prior immutable runner-image digest.

## Out of scope

- The runner does not perform actions described only by markdown or `doIt: false` steps.
- Unrelated roots and repository-sweep guides do not share browser contexts.
- Standalone milestone execution remains the signal for independent recoverability.

Co-Authored-By: Warp <agent@warp.dev>

### CODE REVIEW

**Review by moxious — CHANGES_REQUESTED** [review]
## Merge contract

Fix this item and this PR is mergeable.

1. [blocking] **e2e_shared_result_id_mismatch — Valid milestone results are discarded when planner and guide IDs differ** (medium · correctness-and-reliability)
   The shared runner publishes `TestResultsData.guide.id` from the guide JSON, but `parsePartialChainResults` accepts a result only when that ID equals the repository/planner ID. Local path and journey resolution accepts content whose valid JSON ID differs from its repository key, so Playwright can pass and the CLI will still discard that result and every later result as `REPORT_MISSING`; isolated execution at the base had no such equality gate.
   Required: Make shared execution publish the planned chain ID, or reject planner/content ID mismatches before spawning Playwright, and add a regression test for an accepted local path whose repository key differs from the content ID.

## Other hypotheses checked and cleared

- Bearer tokens are attached only to requests matching the target origin.
- Tokens stay out of the serialized chain input, and tracing is disabled for bearer-token runs.
- Target URLs require HTTP(S); milestone locations must remain same-origin.
- Network, server, and browser loss remain distinct from authentication expiry.
- A nonzero Playwright exit cannot be hidden by already-written passing reports.
- Partial results survive process termination, apart from the confirmed ID mismatch.
- Browser termination monitoring uses bounded cleanup.
- Later milestone recovery does not reload the Grafana page or erase unsaved application state.
- The public report remains schema version `1.0.0`, with one ordered report per milestone.
- The new internal chain schema rejects invalid or duplicate IDs, self-dependencies, unresolved dependencies, unsupported protocols, and unknown root fields.
- Shared-session state remains runtime-only and reversible.
- The CLI/E2E and guide-schema contract-evolution checks both found coherent extensions with no competing owners.
- The hypothesis that terminal session loss should produce `SKIPPED_PREREQ` for dependent milestones was independently refuted: the new contract intentionally assigns zero-step terminal results to every unrun milestone after authentication or browser loss.



PR Review: https://github.com/grafana/grafana-pathfinder-app/pull/1728
Summary: preserve browser state across journey milestones
Verdict: Request Changes
Results: 1 blocker, 0 non-blocking findings, 0 follow-ups

File: tests/e2e-runner/utils/guide-runner/run-guide.ts [hunk 1]
```diff
@@ -0,0 +1,219 @@
+import type { Page } from '@playwright/test';
+
+import { testIds } from '../../../../src/constants/testIds';
+import { contentDigest, type TestResultsData } from '../../../../src/cli/e2e/e2e-reporter';
+import { printDetailedSummary, printDiscoveryResults, printHeader, printStepResult } from '../console-reporter';
+import { dismissBadgeCelebrations } from './badge-celebrations';
```
[the remark this answers is not in the document]
Comment (leslie-heinzen) [reply tests/e2e-runner/utils/guide-runner/run-guide.ts]: Removed the unused `dismissBadgeCelebrations` import in 87e5c603.

**Comment by leslie-heinzen** [comment]
Addressed the blocking review in 87e5c603.

- Shared runs now report the planned chain ID. Standalone runs still use the guide content ID.
- The local path fixture now uses different repository and content IDs. Its unit and runtime checks require planned report IDs.
- Removed the unused `dismissBadgeCelebrations` import.

Validation:
- 34 focused tests passed.
- TypeScript, ESLint, Prettier, Go tests, coverage, review-contract, and script behavior tests passed.
- The full gate could not run `lint:go` because `golangci-lint` is not installed locally. This change does not touch Go.
- Script tests skipped ShellCheck because it is not installed locally.

File: tests/e2e-runner/utils/shared-chain.ts [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+): MilestoneTransition {
+  const currentPath = sameOriginPath(targetUrl, currentUrl);
+  if (authoredStartingLocation === undefined && !isFirstRunnable) {
+    return { startingLocation: currentPath, navigateToStartingLocation: false };
+  }
+  const startingLocation = sameOriginPath(targetUrl, authoredStartingLocation ?? '/');
+  return {
+    startingLocation,
+    navigateToStartingLocation: isFirstRunnable || currentPath !== startingLocation,
+  };
```
[the remark this answers is not in the document]
Comment (leslie-heinzen) [reply tests/e2e-runner/utils/shared-chain.ts:53]: Confirmed and fixed in 612f939a. The first transition no longer reads `about:blank`. It validates the starting location and navigates directly. A regression test covers this path.

File: tests/e2e-runner/utils/shared-chain.ts [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    if (!endsSharedSession(result)) {
+      continue;
+    }
+
+    authExpired = result.abortReason === 'AUTH_EXPIRED' || result.errorCode === 'AUTH_EXPIRED';
+    for (const unrun of input.guides.slice(index + 1)) {
+      results.push(unrunSharedSessionResult(unrun, input.targetUrl, result));
+      executor.publish(results);
+    }
+    break;
```
[the remark this answers is not in the document]
Comment (leslie-heinzen) [reply tests/e2e-runner/utils/shared-chain.ts:184]: Confirmed and fixed in 612f939a. A thrown guide error now continues when the browser session remains available. Browser crashes and closed pages or contexts still stop the chain. Regression tests cover both outcomes.

File: tests/e2e-runner/shared-guide-runner.spec.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+              ? join(artifactsDir, `milestone-${String(index + 1).padStart(3, '0')}-${guide.id}`)
+              : undefined;
+            return runGuideOnPage(page, parsePageGuide(guide.path, guide.content), {
+              targetUrl: chainInput.targetUrl,
+              startingLocation: transition.startingLocation,
+              navigateToStartingLocation: transition.navigateToStartingLocation,
+              replacePreviousGuide: index > 0,
+              previousGuideHadInteractiveSteps:
+                index > 0 && countInteractiveBlocks(JSON.parse(chainInput.guides[index - 1]!.content)) > 0,
+              allowReloadRecovery: index === 0,
```
[the remark this answers is not in the document]
Comment (leslie-heinzen) [reply tests/e2e-runner/shared-guide-runner.spec.ts:114]: Confirmed as a real edge case, but it is medium severity. It requires a skipped milestone between two executed milestones and different interactive states. I left it unchanged under the high-severity-only scope.

**Review by moxious — CHANGES_REQUESTED** [review]
## Merge contract

Fix this item and this PR is mergeable.

1. [blocking] **e2e_shared_recoverable_load_failure_cascade — Recoverable load errors poison later independent milestones** (medium · cli-and-e2e-runner)
   runSharedGuideChain now continues any thrown runGuide failure while the browser is open, but the shared adapter still sets replacePreviousGuide from index > 0. If page.goto, the Help/panel waits, or guide opening fails before a bundled:e2e-test tab becomes active, the next independent milestone immediately fails replacement against the absent tab and the REPORT_MISSING cascade repeats instead of running.
   Required: Carry whether a guide tab actually opened, or inspect the active E2E tab before replacement, and add an adapter-level regression where the first guide fails before tab activation and a later independent guide still runs.



PR Review: https://github.com/grafana/grafana-pathfinder-app/pull/1728
Summary: preserve browser state across journey milestones
Verdict: Request Changes
Results: 1 blocker, 0 non-blocking findings, 0 follow-ups

**Comment by leslie-heinzen** [comment]
Addressed `e2e_shared_recoverable_load_failure_cascade` in cb7547ba.

Replacement now returns when no E2E guide tab is active. A later independent milestone can then open normally. Active E2E tabs still use the existing reset and close checks.

Added an adapter regression for failure before tab activation. Validation passed: 21 focused tests, typecheck, Prettier, and diff checks.

**Review by moxious — CHANGES_REQUESTED** [review]
## Merge contract

Fix this item and this PR is mergeable.

1. [blocking] **e2e_shared_skipped_replacement_leaves_prior_guide_state — The runner skips the guide reset when the old guide tab is open but not focused** (medium · cli-and-e2e-runner)
   Before this commit, the runner stopped with an error when the focused docs tab was not the test guide. It now returns and does nothing. This test is the first action of loadGuide. It runs before page.goto, before the Help wait, and before the docs panel opens. The plugin writes window.__DocsPluginActiveTabUrl only while the docs panel is mounted, and a full page load erases the value. A guided step can cause a full page load. The runner already handles that case at execution.ts:757-769. In this state the value is empty, but the old guide tab comes back from tab storage and the old step progress is still on disk. The runner then skips two actions: the Reset guide click and the tab close. Two faults follow. Fault one is a false pass. Every milestone shares the content key bundled:e2e-test. deriveStepId hashes only the section, the index, the action, the target, and the variant, so the old completions attach to the new milestone steps. The runner marks those steps as skipped. summarizeResults treats a run of only skipped steps as a pass. The milestone reports passed and runs no step at all. Fault two is a late failure. openDocsPage never reuses a tab with the same URL, so a second bundled:e2e-test tab opens. When a later milestone closes one tab, focus moves to the other one. The final wait for the active URL to change can then never succeed. It uses all 15 seconds and fails. The report blames a milestone that is not at fault.
   Required: Do not use the focused tab as the test. Record whether the previous milestone opened a guide tab, or look for any tab with the URL bundled:e2e-test. On the skip path, erase the step progress for bundled:e2e-test. Add a test where an old guide tab is still open and the browser globals are empty. Note: this record also corrects the skipped-prerequisite case that Bugbot reported on this PR.

## Follow-ups

These are tracked separately and do not block merge.

1. [follow_up] **e2e_replacement_contract_doc_omits_skip_branch — The new contract document does not describe the new skip** (low · cli-and-e2e-runner)
   This PR adds the runner guide-load contract to E2E_TESTING_CONTRACT.md. That section lists 11 replacement steps and one exception, for a guide that has no steps. The new code adds a second exception, and the document does not mention it. The same section says to update the runner specs, the contract tests, and the document in one change. This PR also moves the cli-and-e2e-runner contract anchor to #1728, and that anchor does not mention the exception.
   Follow-up: Add the new exception and its condition to the replacement sequence in E2E_TESTING_CONTRACT.md. Add the same condition to the cli-and-e2e-runner contract anchor. Make both edits in this PR.

2. [follow_up] **e2e_pre_activation_regression_not_at_adapter_level — The new test does not cover the chain or the value the adapter sends** (low · testing-and-verification)
   The new test calls replacePreviousE2EGuide and openLegacyE2EGuide directly. It does not run loadGuide, runGuideOnPage, or runSharedGuideChain. It therefore proves the early return, but not that a later milestone still runs. The test also sends previousGuideHadInteractiveSteps as false. The real adapter sends true after a milestone that has interactive steps. Today both values take the same path, because the URL test comes first. No test holds that order. A change to the order would bring the fault back, and the suite would stay green.
   Follow-up: Add an early-return case with previousGuideHadInteractiveSteps set to true. Add a chain-level case where the first milestone fails before its guide tab opens. Assert that a later independent milestone gives a result other than REPORT_MISSING.



PR Review: https://github.com/grafana/grafana-pathfinder-app/pull/1728
Summary: preserve browser state across journey milestones
Verdict: Request Changes
Results: 1 blocker, 0 non-blocking findings, 2 follow-ups

**Comment by leslie-heinzen** [comment]
Addressed `e2e_shared_skipped_replacement_leaves_prior_guide_state` in cae291fe.

- The shared adapter records the exact opened E2E tab ID and its interactive state.
- A later milestone restores the panel, reactivates that visible or overflowed tab, resets progress, and closes the tab.
- If no prior guide tab opened, the runner clears every `bundled:e2e-test` progress namespace before it opens the next guide.
- The adapter-level test covers replacement order and the exact tab ID. The replacement tests cover empty globals and an interactive previous guide.
- The runner contract and review anchor now describe both paths.

Validation passed: 52 focused tests, typecheck, Prettier, diff checks, and 63 review-contract tests.

File: tests/e2e-runner/utils/guide-runner/milestone-replacement.ts [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+            runtimeWindow.__DocsPluginActiveTabId === expectedId &&
+            runtimeWindow.__DocsPluginActiveTabUrl === expectedUrl
+          );
+        },
+        { expectedId: tabId, expectedUrl: E2E_GUIDE_URL },
+        { timeout: REPLACEMENT_TIMEOUT_MS }
+      ),
+      tab.waitFor({ state: 'visible', timeout: REPLACEMENT_TIMEOUT_MS }),
+      overflow.waitFor({ state: 'visible', timeout: REPLACEMENT_TIMEOUT_MS }),
+    ]);
```
[the remark this answers is not in the document]
Comment (leslie-heinzen) [reply tests/e2e-runner/utils/guide-runner/milestone-replacement.ts:92]: Investigated and not reproducible as a correctness failure. `Promise.race` attaches rejection handlers to every input promise. A losing Playwright wait can continue until its timeout, but its later rejection remains handled by the settled race. It cannot fail the shared run or cause an unhandled rejection. A focused Node reproduction also produced no `unhandledRejection`. The remaining cost is temporary polling for at most 15 seconds, which is a low-priority resource concern rather than a blocker.

**Review by moxious — CHANGES_REQUESTED** [review]
## Merge contract

Fix these items and this PR is mergeable.

1. [blocking] **e2e_stale_recorded_tab_after_partial_replacement_cascades — A replacement failure after the tab close poisons the recorded tab ID for every later milestone** (high · cli-and-e2e-runner)
   The new tab-ID record is cleared only on the success path. `loadGuide` calls `onPreviousGuideCleared` after `replacePreviousE2EGuide` resolves (run-guide.ts:67-68), but that function clicks the tab close control at milestone-replacement.ts:259-261 and only then awaits `waitForStepHandlesToDetach` (:266) and a final `waitForFunction` (:267-271). A throw at either point, or a Playwright actionability re-check failing on the close click itself after the node detaches, leaves `previousGuideTab` pointing at a tab that is already gone. shared-guide-runner.spec.ts:117-122 are the only writers of that record and neither `runGuideOnPage` nor `runSharedGuideChain` resets it on failure. The stale ID then makes things worse rather than degrading: because milestone-replacement.ts:227 reads `previousGuideTabId ?? (focused-tab fallback)`, a non-empty stale ID is truthy and bypasses the `clearSharedE2EProgress` skip path entirely. `activateE2EGuideTab` has no recovery for an unknown ID, so it burns the full 15s and throws. shared-chain.ts:169-177 records `infrastructure_error` / `REPORT_MISSING` and continues, and path and journey milestones carry no hard dependencies (guide-chains.ts:127-129; guide-chains.test.ts:42-56 pins `dependencies: []`), so nothing blocks the rest of the journey. Every remaining milestone re-reads the same dead ID, burns another 15s, and reports `REPORT_MISSING` against a milestone that is not at fault. That is the misattributed-failure harm the round-3 blocker named, reached through a different trigger. A reachable live-browser trigger exists: `openDocsPage` never reuses a same-URL tab and mints a fresh `tab-${Date.now()}` ID (docs-panel.tsx:843-873), so two coexisting `bundled:e2e-test` tabs keep `__DocsPluginActiveTabUrl` equal to the E2E URL after the close and time out the wait at :267. Page or browser death is not the trigger, because that makes `browserSessionEnded()` true and the chain breaks cleanly instead.
   Required: Drop the recorded tab ID as soon as the close click succeeds, or invoke `onPreviousGuideCleared` from a `finally`. Separately, treat a recorded tab that is no longer present as the skip path rather than a hard error, so one bad transition cannot poison the rest of the chain.

2. [blocking] **e2e_reset_control_probed_before_restored_guide_renders — Moving replacement after navigation probes the Reset guide control before the restored tab finishes loading** (medium · cli-and-e2e-runner)
   This commit moved `replacePreviousE2EGuide` from the first statement of `loadGuide` to after `page.goto`, the Help wait, and a new `ensureDocsPanelOpen` (run-guide.ts:60-69). After a real navigation the docs panel restores its tab strip with no content: tab-storage-restore.ts always emits `content: null` and docs-panel.tsx:269-288 fires `loadTab(..., 'browser_restore')` afterwards. The `Reset guide` control lives behind that content gate, `activeTab?.content && !activeTab.isLoading` at DocsPanelContentArea.tsx:249. The runner's readiness signals all fire earlier: `useGlobalActiveTabExposure` publishes both window globals in a layout effect at strip render, `ensureDocsPanelOpen` waits only for the panel container (bootstrap.ts:112-161), and in the restore case `activateE2EGuideTab` is skipped entirely because the restored ID and URL already match. milestone-replacement.ts:243 then reads `resetButton.count()` exactly once with no `waitFor` and no retry, and :251-253 throws `'The interactive E2E guide has no Reset guide control'` when the count is 0 and the previous guide had interactive steps. The only slack is `dismissBadgeCelebrations`, roughly 100ms of polling when no toast exists. The asymmetry is the tell: the post-open loading-state wait next door gets 15s (run-guide.ts:73) and this probe gets none. A healthy guide can therefore intermittently report `REPORT_MISSING`, and because the shared spec is a single Playwright test the whole journey fails with it. The pre-`cae291fe` ordering probed a still-mounted, fully loaded panel and had no such exposure. Two honest narrowings: the path needs `navigateToStartingLocation` true, which `resolveMilestoneTransition` (shared-chain.ts:36-54) produces only when an authored starting location differs from the current path; and the PR's own fixture omits that on its second milestone, so the shipped e2e coverage never exercises this. Authored packages do routinely declare per-guide starting locations, which is exactly the 11-of-18 non-root case the PR body cites as its motivation.
   Required: Wait for the Reset guide control, or for `docs-panel-loading-state` to be hidden, before probing its count — guarded by the interactive-blocks flag so a genuine zero-step guide still takes the direct-close path.

## Follow-ups

These are tracked separately and do not block merge.

1. [follow_up] **e2e_tab_reactivation_paths_unverified — The reactivation branches the fix exists for are covered by no test** (medium · testing-and-verification)
   The round-3 blocker required a test with an old guide tab still open and the browser globals empty. The new case at milestone-replacement.test.ts:173-182 deletes both globals but the harness mocks `tabButton.count` to resolve 1 (:63-70), so `tab.count() === 0` is false and it takes the already-rendered-strip branch. The state it claims to model cannot occur that way in a browser: the tab DOM node and the globals are produced in the same React commit, so empty globals imply the tab element is absent. The branches that actually run then — the `Promise.race` at :75-92, the `restoredActive` early return at :93-96, and the overflow-dropdown path at :100-106 — have no coverage, and the harness could not even run them, since `overflow` resolves to the close-button mock which has no `count` or `waitFor`. Nothing covers `replacePreviousE2EGuide` rejecting after the close. The tab-tracking closure that the whole fix consists of (shared-guide-runner.spec.ts:105-122) is unreachable from jest: `jest.config.js:33-36` covers `tests/e2e-runner/utils/**/*.test.ts` only, and that file is a Playwright spec needing a live Grafana. `run-guide.test.ts` does genuinely pin the new `loadGuide` ordering and the `'old-tab'` hand-off, but with `countInteractiveBlocks` mocked to 0 it takes the early return at run-guide.ts:185-196 that hardcodes `passed`, so its outcome assertion cannot fail.
   Follow-up: Add a harness case that starts with `tab.count()` at 0 and gives `overflow` a real mock, so the race and the dropdown path run; add a case where `replacePreviousE2EGuide` rejects after the close; and extract the tab-tracking closure out of the Playwright spec so jest can assert the recorded ID across a failed milestone.



PR Review: https://github.com/grafana/grafana-pathfinder-app/pull/1728
Summary: preserve browser state across journey milestones
Verdict: Request Changes
Results: 2 blockers, 0 non-blocking findings, 1 follow-up

**Review by moxious — DISMISSED** [review]
**Blocker 1 — stale tab ID cascade.** When `replacePreviousE2EGuide` throws *after* it has already closed the previous tab, the recorded tab ID is never cleared, and because a non-empty stale ID is truthy it bypasses the graceful skip path instead of falling back to it — so every remaining milestone burns 15s and reports `REPORT_MISSING` against a milestone that is innocent, turning one bad transition into a whole journey of misattributed failures. This needs a prior fault to trigger it (the realistic route is two coexisting `bundled:e2e-test` tabs), so it will not fire on a healthy run: **genuine edge case, low likelihood, safe as a follow-up.**

**Blocker 2 — reset probe races the restored tab.** On any later milestone that navigates, the tab strip restores with `content: null` and the `Reset guide` control does not exist yet, but the runner's readiness signals (window globals in a layout effect, panel-container visibility) all fire earlier, and then `resetButton.count()` is read exactly once with no `waitFor` — a zero throws and fails the whole single-test journey. The trigger is a milestone whose authored starting location differs from the current path, which is the *common* case in real journeys — the PR body itself cites 11 of 18 failures having non-root starting locations — and the only slack is ~100ms of badge polling, so this is a flaky hard failure on healthy guides, not an edge case. Critically, **the PR's own fixture omits an authored starting location on its second milestone, so green CI is not evidence against this.**

**Recommendation: fix 2, merge, file 1.** Blocker 2 is the only one that threatens proper functioning out of the gate, and it's a one-line `waitFor`. That the author operates the infrastructure cuts in favour of merging blocker 1 as a follow-up — but it cuts *against* shipping blocker 2, because a runner that intermittently blames healthy guides corrupts exactly the guide-health signal he'd be reading to notice the problem.

**Comment by leslie-heinzen** [comment]
Addressed the restored-tab reset race in 29c1d45c.

The shared runner now waits for the restored interactive guide's `Reset guide` control before it captures prior step handles. The zero-step path still closes directly without this wait.

The regression test models a restored tab whose ID and URL are available before its Reset control renders. It also models the persistent missing-control case as a Playwright timeout.

Validation passed:
- 19 focused runner tests
- `npm run typecheck`
- `npm run lint`
- `npm run prettier-test`
- `git diff --check`

This commit is limited to blocker 2. The stale tab ID cascade remains separate follow-up work, as recommended.

**Review by Jayclifford345 — APPROVED** [review]
